### ISSUE TITLE
Inconsistent detections after custom training

### ISSUE THREAD

**Issue by kaminetzky** [ultralytics/yolov5#8468]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hey there!

I'm training a YOLOv5 network for detecting wrenches in X-ray images of baggage. I'm training YOLOv5X with thousands of images for 300 epochs. Training images contain simulated wrenches while testing images contain real wrenches. All parameters are the default ones except for the batch size (set to 36 for 3 GPUs).

I noticed in my last run that tiny differences between images result in considerably different predictions. For example, the viewing angle changes slightly or the image is cropped differently. Here are some examples:

![P05561](https://user-images.githubusercontent.com/11022568/177195376-625c0744-e98d-41c8-be00-d0ee674a4e81.jpg)
![P05562](https://user-images.githubusercontent.com/11022568/177195381-918689e4-e163-4dfd-b2aa-b852137b1ff7.jpg)

![P01297](https://user-images.githubusercontent.com/11022568/177195469-aad58941-497d-4150-abf6-5c2063d43f20.jpg)
![P01298](https://user-images.githubusercontent.com/11022568/177195492-d60c08d8-8cec-4fc9-97d7-909f66b1b6fe.jpg)

Has anybody faced a similar issue or know how to fix it? Any help is appreciated.

Thanks!

**Comment by glenn-jocher** [comment]
@kaminetzky 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by kaminetzky** [comment]
Hey Glenn! Thanks for the quick reply. I'm following all YOLO training tips. I'll add the requested data ASAP.

I just find it odd that tiny changes between the images result in significant differences in the predictions. I thought that maybe somebody has seen something like this.

**Comment by kaminetzky** [comment]
Important: the training and validation subsets contain **simulated** wrenches, while the testing subset contains **real** wrenches. This is why validation and testing metrics are very different.

Testing metrics:

               Class     Images     Labels          P          R     mAP@.5   mAP@0.5:0.95
                 all       2191        878      0.403      0.383      0.451          0.352

Training metrics:

<img width="1633" alt="image" src="https://user-images.githubusercontent.com/11022568/177200096-aaba1f61-052b-4780-a0e8-5d78c89bc4aa.png">

<img width="1634" alt="image" src="https://user-images.githubusercontent.com/11022568/177200111-00d983f8-786b-44b3-a8ed-438cfe8fde9a.png">

Validation (not testing) confusion matrix:
![confusion_matrix](https://user-images.githubusercontent.com/11022568/177201066-58e9d65b-4cce-4165-92c8-12fbfb659046.png)

Label stats:

![labels](https://user-images.githubusercontent.com/11022568/177201159-d80a38fe-d6d4-44f1-affc-a2af0ed5b4e6.jpg)
![labels_correlogram](https://user-images.githubusercontent.com/11022568/177201171-1ca89fcd-dfd9-464f-9ad8-cefc6d6f3cd0.jpg)

PR curve:

![PR_curve](https://user-images.githubusercontent.com/11022568/177201189-0b7fab66-1552-47d4-b52b-62a18af7aaee.png)


Training mosaic:

![train_batch0](https://user-images.githubusercontent.com/11022568/177201211-c82ea915-bc6f-48f7-a42d-e07e1b171485.jpg)

Validation labels and predictions:

![val_batch0_labels](https://user-images.githubusercontent.com/11022568/177201269-ea730668-a8eb-4e8f-96bf-049a0ed4a78d.jpg)

![val_batch0_pred](https://user-images.githubusercontent.com/11022568/177201276-bb0caa9b-e670-4a8e-8841-d5eca750803b.jpg)

**Comment by glenn-jocher** [comment]
@kaminetzky for best results your train/val/test/deployed images should originate from a shared image space. If you use simulated data to train I would not expect good results unless you apply the model to similarly simulated images.

<img width="2029" alt="dataset" src="https://user-images.githubusercontent.com/26833433/177201767-979893f5-a81d-467e-a236-dd2e253e5341.png">

**Comment by kaminetzky** [comment]
Thanks for the info, Glenn!

The goal of my investigation is to generate simulated threat objects that are similar to the real ones. This would allow the model to be trained when real object data is not available.

Hopefully, somebody has faced a similar issue. :)

**Comment by glenn-jocher** [comment]
This is a common discussion in the ML community, where you can find arguments both ways i.e. Waymo with simulated data vs Tesla with real world data. 

In my experience Tesla has the winning approach here.

**Comment by kaminetzky** [comment]
Gotcha. Thanks, Glenn! :)

**Comment by huihuid** [comment]
> @kaminetzky for best results your train/val/test/deployed images should originate from a shared image space. If you use simulated data to train I would not expect good results unless you apply the model to similarly simulated images.
> 
> <img alt="dataset" width="2000" src="https://user-images.githubusercontent.com/26833433/177201767-979893f5-a81d-467e-a236-dd2e253e5341.png">

Hi @glenn-jocher, really like this slide you shared about dataset problem above, it looks like one page in slides, may I ask for the whole slides? BTW, thanks for your and the team's amazing work in Yolo!

**Comment by kaminetzky** [comment]
Hey @huihuid! I'm aware that the images should come from the same shared space. In fact, some past experiments where the training set included simulated and real images went great! However, my goal now is to be able to train the model for objects where no data is available. For this, I'm trying to generate sim images that look as the real ones. Cheers!

**Comment by glenn-jocher** [comment]
@huihuid sorry I don't have the original deck anymore, just a few slides we saved. We've got a recent talk on YouTube though from WAICF a few months ago that shows some cool YOLOv5 use cases and ideas.

https://youtu.be/iSXPMH3Wfig

**Comment by CongSuxu** [comment]
@kaminetzky I face the same problem recently, have ever find a solution?

**Comment by kaminetzky** [comment]
Hi @CongSuxu, I was not able to find a solution to this problem.

**Comment by CongSuxu** [comment]
I am encountering a situation where two images that appear identical to the naked eye produce different results when processed using YOLOv8L.

In both images, the target is detected with a confidence score of 0.9869. However, in one image, the block is also detected with a confidence score of 0.7459, whereas in the other image, the block is not detected at all. If the block was detected but filtered out by NMS, its confidence score must have been below 0.01.

Attached below are the detection results for both images, as well as the result of their direct subtraction for reference.
![2](https://github.com/user-attachments/assets/81360180-3651-4e0c-98c6-82a1ca2f8ce4)
![3](https://github.com/user-attachments/assets/16f0e264-a99b-434d-96e5-298dcbe2c0b7)
![0](https://github.com/user-attachments/assets/bf106c7e-dfd6-4727-a162-5d90ba17df37)
I can provide the dataset, the trained model, and the test images if anyone is interested. The train set has 2300 images and val set has 575 images.

**Comment by pderrenger** [comment]
Thank you for sharing the detailed information and results. The difference in detections could be due to subtle variations in pixel values or artifacts that are imperceptible to the human eye but can influence the model's output. This sensitivity may arise from:

1. **Training and Dataset Composition**: Ensure your dataset is diverse and representative of the test environment. Minor inconsistencies can lead to variations in predictions, especially for edge cases.
   
2. **Confidence Thresholds and NMS**: The block not being detected might indicate it was filtered out due to Non-Maximum Suppression (NMS) or confidence thresholds. You can adjust these parameters during inference to investigate further.

If the issue persists, consider visualizing the feature maps or intermediate layers to understand how the model interprets these images. Additionally, testing with the latest YOLOv8 version from [Ultralytics](https://github.com/ultralytics/ultralytics) is recommended to ensure compatibility and incorporate recent improvements. Let me know if you need further assistance!
